Also, some new test failures:

<pre>
======================================================================
ERROR: test_delete (miro.test.downloadertest.DownloaderTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/z3p/Projects/miro/tv/linux/dist/usr/lib/python2.7/site-packages/miro/test/framework.py", line 204, in _uses_httpclient
    return fun(*args, **kwargs)
  File "/home/z3p/Projects/miro/tv/linux/dist/usr/lib/python2.7/site-packages/miro/test/downloadertest.py", line 65, in test_delete
    self.setup_state()
  File "/home/z3p/Projects/miro/tv/linux/dist/usr/lib/python2.7/site-packages/miro/test/downloadertest.py", line 24, in setup_state
    downloader.init_controller()
AttributeError: 'module' object has no attribute 'init_controller'

======================================================================
ERROR: test_delete (miro.test.downloadertest.DownloaderTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/z3p/Projects/miro/tv/linux/dist/usr/lib/python2.7/site-packages/miro/test/downloadertest.py", line 28, in tearDown
    downloader.shutdown_downloader(
AttributeError: 'module' object has no attribute 'shutdown_downloader'

======================================================================
ERROR: test_download (miro.test.downloadertest.DownloaderTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/z3p/Projects/miro/tv/linux/dist/usr/lib/python2.7/site-packages/miro/test/framework.py", line 204, in _uses_httpclient
    return fun(*args, **kwargs)
  File "/home/z3p/Projects/miro/tv/linux/dist/usr/lib/python2.7/site-packages/miro/test/downloadertest.py", line 60, in test_download
    self.setup_state()
  File "/home/z3p/Projects/miro/tv/linux/dist/usr/lib/python2.7/site-packages/miro/test/downloadertest.py", line 24, in setup_state
    downloader.init_controller()
AttributeError: 'module' object has no attribute 'init_controller'

======================================================================
ERROR: test_download (miro.test.downloadertest.DownloaderTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/z3p/Projects/miro/tv/linux/dist/usr/lib/python2.7/site-packages/miro/test/downloadertest.py", line 28, in tearDown
    downloader.shutdown_downloader(
AttributeError: 'module' object has no attribute 'shutdown_downloader'

======================================================================
ERROR: test_remove_before_downloader_referenced (miro.test.itemtest.ItemRemoveTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/z3p/Projects/miro/tv/linux/dist/usr/lib/python2.7/site-packages/miro/test/itemtest.py", line 189, in test_remove_before_downloader_referenced
    item.remove()
  File "/home/z3p/Projects/miro/tv/linux/dist/usr/lib/python2.7/site-packages/miro/item.py", line 1944, in remove
    self.set_downloader(None)
  File "/home/z3p/Projects/miro/tv/linux/dist/usr/lib/python2.7/site-packages/miro/item.py", line 1890, in set_downloader
    self.downloader.remove_item(self)
  File "/home/z3p/Projects/miro/tv/linux/dist/usr/lib/python2.7/site-packages/miro/downloader.py", line 745, in remove_item
    self.remove()
  File "/home/z3p/Projects/miro/tv/linux/dist/usr/lib/python2.7/site-packages/miro/downloader.py", line 716, in remove
    app.download_state_manager.total_down_rate -= rates[0]
AttributeError: 'NoneType' object has no attribute 'total_down_rate'
</pre>

--- a/tv/lib/app.py
+++ b/tv/lib/app.py
@@ -74,6 +74,9 @@ backend_config_watcher = None
 frontend_config_watcher = None
 downloader_config_watcher = None
 
+# download state manager
+download_state_manager = None
+
 # sends MetadataProgressUpdate messages to the frontend
 metadata_progress_updater = None
 

--- a/tv/lib/controller.py
+++ b/tv/lib/controller.py
@@ -69,7 +69,9 @@ class Controller:
         logging.info("Shutting down video conversions manager")
         conversions.conversion_manager.shutdown()
         logging.info("Shutting down Downloader...")
-        downloader.shutdown_downloader(self.downloader_shutdown)
+        if app.download_state_manager is not None:
+            app.download_state_manager.shutdown_downloader(
+                self.downloader_shutdown)
         try:
             logging.info("Shutting down worker process.")
             workerprocess.shutdown()

--- a/tv/lib/dl_daemon/command.py
+++ b/tv/lib/dl_daemon/command.py
@@ -109,34 +109,17 @@ class RemoveHTTPAuthCommand(Command):
         from miro import httpauth
         httpauth.remove_by_url_and_realm(*self.args)
 
-class UpdateDownloadStatus(Command):
-    def action(self):
-        from miro.downloader import RemoteDownloader
-        return RemoteDownloader.update_status(*self.args, **self.kws)
-
-# DownloaderSyncRequest/DownloaderSyncNotify: what are these?
-#
-# They are used to as a notification.  Usage: DownloaderSyncRequest command
-# is sent to the downloader and when it is received the downloader sends
-# DownloaderSyncNotify.  Thus, when a DownloaderSyncNotify is received,
-# you can be sure that all commands for the corresponding DownloaderSyncRequest
-# has been completed (commands are never re-ordered).
-class DownloaderSyncRequest(Command):
-    def action(self):
-        from miro.dl_daemon import download
-        return download.sync_notify()
-
-class DownloaderSyncNotify(Command):
-    def action(self):
-        from miro.messages import DownloaderSyncCommandComplete
-        DownloaderSyncCommandComplete().send_to_frontend()
-
 class BatchUpdateDownloadStatus(Command):
     spammy = True
     def action(self):
         from miro.downloader import RemoteDownloader
-        for status in self.args[0]:
-            RemoteDownloader.update_status(status)
+        from miro.messages import DownloaderSyncCommandComplete
+
+        cmd_done = self.args[1]
+        fresh = all(RemoteDownloader.update_status(status, cmd_done=cmd_done)
+                    for status in self.args[0])
+        if cmd_done and fresh:
+            DownloaderSyncCommandComplete().send_to_frontend()
 
 class DownloaderErrorCommand(Command):
     def action(self):
@@ -171,6 +154,9 @@ class ShutDownResponseCommand(Command):
 #############################################################################
 #  App to Downloader commands                                               #
 #############################################################################
+
+# XXX why do we have so much junk here when we can multiplex this stuff with
+# a cmd parameter? -gl
 class InitialConfigCommand(Command):
     def action(self):
         app.config.set_dictionary(*self.args, **self.kws)
@@ -186,46 +172,63 @@ class UpdateHTTPPasswordsCommand(Command):
         from miro.dl_daemon.private import httpauth
         httpauth.update_passwords(*self.args)
 
-class StartNewDownloadCommand(Command):
+# Downloader Daemon start/stop/resume demux.
+#
+# This is the class that contains the action handler for commands which may
+# be marshalled and sent to us via the DownloadStateManager().
+#
+# Here, they are demuxed, and then dispatched to the downloader.
+class DownloaderBatchCommand(Command):
+    STOP    = 0
+    RESUME  = 1
+    PAUSE   = 2
+    RESTORE = 4
+
+    RESUME_EXISTING = 0
+    RESUME_NEW      = 1
+
     def action(self):
         from miro.dl_daemon import download
-        return download.start_new_download(*self.args, **self.kws)
-
-class StartDownloadCommand(Command):
-    def action(self):
-        from miro.dl_daemon import download
-        return download.start_download(*self.args, **self.kws)
-
-class PauseDownloadCommand(Command):
-    def action(self):
-        from miro.dl_daemon import download
-        return download.pause_download(*self.args, **self.kws)
-
-class StopDownloadCommand(Command):
-    def action(self):
-        from miro.dl_daemon import download
-        return download.stop_download(*self.args, **self.kws)
-
-class StopUploadCommand(Command):
-    def action(self):
-        from miro.dl_daemon import download
-        return download.stop_upload(*self.args, **self.kws)
-
-class PauseUploadCommand(Command):
-    def action(self):
-        from miro.dl_daemon import download
-        return download.pause_upload(*self.args, **self.kws)
-
-class GetDownloadStatusCommand(Command):
-    def action(self):
-        from miro.dl_daemon import download
-        return download.get_download_status(*self.args, **self.kws)
-
-class RestoreDownloaderCommand(Command):
-    def action(self):
-        from miro.dl_daemon import download
-        return download.restore_downloader(*self.args, **self.kws)
-
+        mark_reply = True
+        for dlid, (cmd, args) in self.args[0].iteritems():
+            if cmd == self.PAUSE:
+                upload = args['upload']
+                if upload:
+                    download.pause_upload(dlid)
+                else:
+                    download.pause_download(dlid)
+            elif cmd == self.STOP:
+                upload = args['upload']
+                if upload:
+                    download.stop_upload(dlid)
+                else:
+                    download.stop_download(dlid, args['delete'])
+            elif cmd == self.RESUME:
+                subcmd = args['subcmd']
+                if subcmd == self.RESUME_EXISTING:
+                    download.start_download(dlid)
+                elif cmd == self.RESUME_NEW:
+                    channel_name = args['channel_name']
+                    url = args['url']
+                    content_type = args['content_type']
+                    download.start_new_download(url, dlid, content_type,
+                                                channel_name)
+                else:
+                    raise ValueError('Invalid resume subcommand')
+            elif cmd == self.RESTORE:
+                # Restoring a downloader doesn't actually change any state
+                # so don't reply.
+                mark_reply = False
+                downloader = args['downloader']
+                download.restore_downloader(downloader)
+            else:
+                raise ValueError('unknown downloader batch command %s' % cmd)
+        # Mark this so that the next time we run through the periodic update
+        # which will be after all the above have been processed because we
+        # are in the same thread.
+        if mark_reply:
+            download.DOWNLOAD_UPDATER.set_cmds_done()
+ 
 class MigrateDownloadCommand(Command):
     def action(self):
         from miro.dl_daemon import download

--- a/tv/lib/dl_daemon/download.py
+++ b/tv/lib/dl_daemon/download.py
@@ -134,7 +134,6 @@ def stop_download(dlid, delete):
         return True
     finally:
         _lock.release()
-
     return download.stop(delete)
 
 def stop_upload(dlid):
@@ -150,9 +149,6 @@ def stop_upload(dlid):
     finally:
         _lock.release()
     return download.stop_upload()
-
-def sync_notify():
-    return command.DownloaderSyncNotify(daemon.LAST_DAEMON).send()
 
 def pause_upload(dlid):
     _lock.acquire()
@@ -200,6 +196,9 @@ def shutdown():
         _downloads[dlid].shutdown()
     logging.info("Shutting down torrent session...")
     TORRENT_SESSION.shutdown()
+    # Flush the status updates.
+    logging.info('flushing status updates...')
+    DOWNLOAD_UPDATER.flush_update()
     logging.info("shutdown() finished")
 
 def restore_downloader(downloader):
@@ -424,24 +423,35 @@ class DownloadStatusUpdater(object):
 
     def __init__(self):
         self.to_update = set()
+        self.cmds_done = False
 
     def start_updates(self):
         eventloop.add_timeout(self.UPDATE_CLIENT_INTERVAL, self.do_update,
                 "Download status update")
 
-    def do_update(self):
+    def flush_update(self):
+        self.do_update(periodic=False)
+
+    def do_update(self, periodic=True):
         try:
             TORRENT_SESSION.update_torrents()
             statuses = []
             for downloader in self.to_update:
                 statuses.append(downloader.get_status())
             self.to_update = set()
-            if statuses:
+            if statuses or self.cmds_done:
                 command.BatchUpdateDownloadStatus(daemon.LAST_DAEMON,
-                        statuses).send()
+                                                  statuses,
+                                                  self.cmds_done).send()
+                self.cmds_done = False
         finally:
-            eventloop.add_timeout(self.UPDATE_CLIENT_INTERVAL, self.do_update,
-                    "Download status update")
+            if periodic:
+                eventloop.add_timeout(self.UPDATE_CLIENT_INTERVAL,
+                                      self.do_update,
+                                      "Download status update")
+
+    def set_cmds_done(self):
+        self.cmds_done = True
 
     def queue_update(self, downloader):
         self.to_update.add(downloader)
@@ -498,9 +508,12 @@ class BGDownloader(object):
             'retryCount': self.retryCount,
             'channelName': self.channelName}
 
-    def update_client(self):
-        x = command.UpdateDownloadStatus(daemon.LAST_DAEMON, self.get_status())
-        return x.send()
+    def update_client(self, now=False):
+        if not now:
+            DOWNLOAD_UPDATER.queue_update(self)
+        else:
+            command.BatchUpdateDownloadStatus(daemon.LAST_DAEMON,
+                                              [self.get_status()]).send()
 
     def pick_initial_filename(self, suffix=".part", torrent=False,
                               is_directory=False):
@@ -747,9 +760,7 @@ class HTTPDownloader(BGDownloader):
             self.url, self.on_download_finished, self.on_download_error,
             header_callback=self.on_headers, write_file=self.filename,
             resume=resume)
-        self.update_client()
-        eventloop.add_timeout(self.CHECK_STATS_TIMEOUT, self.update_stats,
-                'update http downloader stats')
+        self.update_stats()
 
     def _resume_sanity_check(self):
         """Do sanity checks to test if we should try HTTP Resume.
@@ -891,7 +902,7 @@ class HTTPDownloader(BGDownloader):
             self.rate = stats.download_rate
         eventloop.add_timeout(self.CHECK_STATS_TIMEOUT, self.update_stats,
                 'update http downloader stats')
-        DOWNLOAD_UPDATER.queue_update(self)
+        self.update_client()
 
     def pause(self):
         """Pauses the download.
@@ -1285,9 +1296,8 @@ class BTDownloader(BGDownloader):
             self.move_to_movies_directory()
             self.state = u"uploading"
             self.endTime = clock()
-            self.update_client()
-        else:
-            DOWNLOAD_UPDATER.queue_update(self)
+
+        self.update_client()
 
         if app.config.get(prefs.LIMIT_UPLOAD_RATIO):
             if status.state == lt.torrent_status.states.seeding:

--- a/tv/lib/downloader.py
+++ b/tv/lib/downloader.py
@@ -31,6 +31,7 @@ import datetime
 import os
 import random
 import logging
+import time
 
 from miro.gtcache import gettext as _
 from miro.database import DDBObject, ObjectNotFoundError
@@ -51,13 +52,157 @@ from miro import flashscraper
 from miro import fileutil
 from miro.fileobject import FilenameType
 
-daemon_starter = None
+class DownloadStateManager(object):
+    """DownloadStateManager: class to store state information about the
+    downloader.
 
-# a hash of download ids that the server knows about.
-_downloads = {}
+    Commands to the downloader is batched and sent every second.  This is
+    based on the premise that commands for a particular download id can 
+    be completely superceded by a subsequent command, with the exception
+    of a pause/resume pair.  For example, a stop command will completely
+    supecede a pause command, so if the 2 are sent in quick succession
+    only the stop command will be sent by the downloader.  The exception
+    to this rule is the pause/resume pair which acts like matter and
+    anti-matter, which will nuke itself when they come into contact
+    (but dies with not even a whimper instead of a gorgeous display).
+    """
+    STOP    = command.DownloaderBatchCommand.STOP
+    RESUME  = command.DownloaderBatchCommand.RESUME
+    PAUSE   = command.DownloaderBatchCommand.PAUSE
+    RESTORE = command.DownloaderBatchCommand.RESTORE
 
-total_up_rate = 0
-total_down_rate = 0
+    RESUME_EXISTING = command.DownloaderBatchCommand.RESUME_EXISTING
+    RESUME_NEW      = command.DownloaderBatchCommand.RESUME_NEW
+
+    UPDATE_INTERVAL = 1
+
+    def __init__(self):
+        self.total_up_rate = 0
+        self.total_down_rate = 0
+        # a hash of download ids that the server knows about.
+        self.downloads = {}
+        self.daemon_starter = None
+        self.startup_commands = dict()
+        self.commands = dict()
+        self.bulk_mode = False
+
+    def set_bulk_mode(self):
+        self.bulk_mode = True
+
+    def send_initial_updates(self):
+        commands = self.startup_commands
+        self.startup_commands = None
+        if commands:
+            c = command.DownloaderBatchCommand(RemoteDownloader.dldaemon,
+                                               commands)
+            c.send()
+
+    def send_updates(self):
+        commands = self.commands
+        self.commands = dict()
+        if commands:
+            c = command.DownloaderBatchCommand(RemoteDownloader.dldaemon,
+                                               commands)
+            c.send()
+        elif self.bulk_mode:
+            from miro.messages import DownloaderSyncCommandComplete
+            # If we did a pause/resume/cancel all, and there weren't any
+            # items in the list to send nobody would re-enable the auto-sort.
+            # So we do it here.
+            DownloaderSyncCommandComplete().send_to_frontend()
+        # Reset the bulk mode notification.
+        self.bulk_mode = False
+        self.start_updates()
+
+    def start_updates(self):
+        eventloop.add_timeout(self.UPDATE_INTERVAL,
+                              self.send_updates,
+                              "Send Download Command Updates")
+
+    def get_download(self, dlid):
+        try:
+            return self.downloads[dlid]
+        except KeyError:
+            return None
+
+    def add_download(self, dlid, downloader):
+        self.downloads[dlid] = downloader
+
+    def delete_download(self, dlid):
+        try:
+            del self.downloads[dlid]
+        except KeyError:
+            return False
+        else:
+            return True
+
+    def daemon_started(self):
+        return self.daemon_starter and self.daemon_starter.started
+
+    def queue(self, identifier, cmd, args):
+        if not self.downloads.has_key(identifier):
+            raise ValueError('add_download() not called before queue()')
+
+        # Catch restores first, we will flush them when the downloader's
+        # started.
+        if cmd == self.RESTORE and not self.daemon_started():
+            self.startup_commands[identifier] = (cmd, args)
+            return
+
+        exists = self.commands.has_key(identifier)
+
+        # Make sure that a pause/resume pair cancel each other out.  For
+        # others, assume that a subsequent command can completely supercede
+        # the previous command.
+        if exists:
+            old_cmd, unused = self.commands[identifier]
+            if (old_cmd == self.RESUME and cmd == self.PAUSE or
+              old_cmd == self.PAUSE and cmd == self.RESUME):
+                # Make sure that we unfreeze it
+                self.downloads[identifier].status_updates_frozen = False
+                del self.commands[identifier]
+                return
+            # HACK: When we pause and resume we currently send a download
+            # command, then a restore downloader command which doesn't
+            # do anything.  This also breaks our general assumption that a 
+            # current command can completely supercede any previous queued
+            # command so if we see it disable it.  I'm not actually
+            # sure why we'd want to send a restore command in this case.
+            if cmd == self.RESTORE:
+               logging.info('not restoring active download')
+               return
+
+        # Freeze the status updates, but don't freeze if it is a restore.
+        if not cmd == self.RESTORE:
+            self.downloads[identifier].status_updates_frozen = True
+        self.commands[identifier] = (cmd, args)
+
+    def init_controller(self):
+        """Intializes the download daemon controller.
+
+        This doesn't actually start up the downloader daemon, that's done
+        in startup_downloader.  Commands will be queued until then.
+        """
+        self.daemon_starter = DownloadDaemonStarter()
+
+    def startup_downloader(self):
+        """Initialize the downloaders.
+    
+        This method currently does 2 things.  It deletes any stale files
+        self in Incomplete Downloads, then it restarts downloads that have
+        been restored from the database.  It must be called before any
+        RemoteDownloader objects get created.
+        """
+        self.daemon_starter.startup()
+        # Now that the daemon has started, we can process updates.
+        self.send_initial_updates()
+        self.start_updates()
+    
+    def shutdown_downloader(self, callback=None):
+        if self.daemon_starter:
+            self.daemon_starter.shutdown(callback)
+        elif callback:
+            callback()
 
 def get_downloader_by_dlid(dlid):
     try:
@@ -72,14 +217,9 @@ def generate_dlid():
         dlid = u"download%08d" % random.randint(0, 99999999)
     return dlid
 
-# NB: sync_downloader_commands: for information on how to use this please
-# see the note in DownloaderSyncRequest/DownloaderSyncNotify.
-def sync_downloader_commands():
-    c = command.DownloaderSyncRequest(RemoteDownloader.dldaemon)
-    c.send()
-
 class RemoteDownloader(DDBObject):
     """Download a file using the downloader daemon."""
+    MIN_STATUS_UPDATE_SPACING = 0.7
     def setup_new(self, url, item, contentType=None, channelName=None):
         check_u(url)
         if contentType:
@@ -105,6 +245,8 @@ class RemoteDownloader(DDBObject):
         self.manualUpload = False
         self._save_later_dc = None
         self._update_retry_time_dc = None
+        self.status_updates_frozen = False
+        self.last_update = time.time()
         if contentType is None:
             self.contentType = u""
         else:
@@ -168,7 +310,7 @@ class RemoteDownloader(DDBObject):
         downloader to disk, do it now.
         """
         if self.id_exists() and self._save_later_dc is not None:
-            self.signal_change()
+            self.signal_change(needs_signal_item=False)
 
     def _cancel_save_later(self):
         if self._save_later_dc is not None:
@@ -230,29 +372,94 @@ class RemoteDownloader(DDBObject):
         return (0, 0)
 
     def before_changing_status(self):
-        global total_down_rate
-        global total_up_rate
         rates = self._get_rates()
-        total_down_rate -= rates[0]
-        total_up_rate -= rates[1]
+        app.download_state_manager.total_down_rate -= rates[0]
+        app.download_state_manager.total_up_rate -= rates[1]
 
     def after_changing_status(self):
-        global total_down_rate
-        global total_up_rate
         self._recalc_state()
         rates = self._get_rates()
-        total_down_rate += rates[0]
-        total_up_rate += rates[1]
+        app.download_state_manager.total_down_rate += rates[0]
+        app.download_state_manager.total_up_rate += rates[1]
 
     @classmethod
-    def update_status(cls, data):
+    def update_status(cls, data, cmd_done=False):
         for field in data:
             if field not in ['filename', 'shortFilename', 'channelName',
                              'metainfo']:
                 data[field] = unicodify(data[field])
+
         self = get_downloader_by_dlid(dlid=data['dlid'])
 
         if self is not None:
+            now = time.time()
+            last_update = self.last_update
+            rate_limit = False
+            state = self.get_state()
+            new_state = data.get('state', u'downloading')
+
+            # If this item was marked as pending update, then any update
+            # which comes in now which does not have cmd_done set is void.
+            if not cmd_done and self.status_updates_frozen:
+                logging.debug('self = %s, '
+                              'saved state = %s '
+                              'downloader state = %s.  '
+                              'Discard.',
+                              self, state, new_state)
+                # treat as stale
+                return False
+
+            # If the timing between the status updates is too narrow,
+            # try to skip it because it makes the UI jerky otherwise.
+            if now < last_update:
+                logging.debug('time.time() gone backwards last = %s now = %s',
+                              last_update, now)
+            else:
+                diff = now - last_update
+                if diff < self.MIN_STATUS_UPDATE_SPACING:
+                    logging.debug('Rate limit: '
+                                  'self = %s, now - last_update = %s, '
+                                  'MIN_STATUS_UPDATE_SPACING = %s.',
+                                  self, diff, self.MIN_STATUS_UPDATE_SPACING)
+                    rate_limit = True
+
+            # If the state is one which we set and was meant to be passed
+            # through to the downloader (valid_states), and the downloader
+            # replied with something that was a response to a previous
+            # download command, and state was also a part of valid_states,
+            # but the saved state and the new state do not match
+            # then it means the message is stale.
+            #
+            # Have a think about why this is true: when you set a state,
+            # which is authoritative, to the downloader you expect it
+            # to reply with that same state.  If they do not match then it
+            # means the message is stale.
+            #
+            # The exception to this rule is if the downloader replies with
+            # an error state, or if downloading has transitioned to finished
+            # state.
+            #
+            # This also does not apply to any state which we set on the
+            # downloader via a restore command.  A restore command before
+            # a pause/resume/cancel will work as intended, and no special
+            # trickery is required.  A restore command which happens after
+            # a pause/resume/cancel is void, so no work is required.
+            #
+            # I hope this makes sense and is clear!
+            valid_states = (u'downloading', u'paused', u'stopped',
+                            u'uploading-paused', u'finished')
+            if (cmd_done and
+              state in valid_states and new_state in valid_states and
+              state != new_state):
+                if not (state == u'downloading' and new_state == u'finished'):
+                    logging.debug('self = %s STALE.  '
+                                  'Saved state %s, got state %s.  Discarding.',
+                                  self, state, new_state)
+                    return False
+
+            # We are updating!  Reset the status_updates_frozen flag.
+            self.status_updates_frozen = False
+ 
             # FIXME - this should get fixed.
             metainfo = data.pop('metainfo', self.metainfo)
 
@@ -262,7 +469,10 @@ class RemoteDownloader(DDBObject):
             current = (self.status, self.metainfo)
             new = (data, metainfo)
             if current == new:
-                return
+                return True
+
+            # We have something to update: update the last updated timestamp.
+            self.last_update = now
 
             was_finished = self.is_finished()
             old_filename = self.get_filename()
@@ -285,7 +495,7 @@ class RemoteDownloader(DDBObject):
             finished = self.is_finished() and not was_finished
             file_migrated = (self.is_finished() and
                              self.get_filename() != old_filename)
-            needs_signal_item = not (finished or file_migrated)
+            needs_signal_item = not (finished or file_migrated or rate_limit)
             self.after_changing_status()
 
             if ((self.get_state() == u'uploading'
@@ -294,21 +504,24 @@ class RemoteDownloader(DDBObject):
                       and self.get_upload_ratio() > app.config.get(prefs.UPLOAD_RATIO)))):
                 self.stop_upload()
 
-            self.signal_change(needs_signal_item=needs_signal_item,
-                               needs_save=False)
             if self.changed_attributes == set(('status',)):
                 # if we just changed status, then we can wait a while
                 # to store things to disk.  Since we go through
                 # update_status() often, this results in a fairly
                 # large performance gain and alleviates #12101
                 self._save_later()
+                self.signal_change(needs_signal_item=needs_signal_item,
+                                   needs_save=False)
             else:
                 self.signal_change()
+
             if finished:
                 for item in self.item_list:
                     item.on_download_finished()
             elif file_migrated:
                 self._file_migrated(old_filename)
+
+        return True
 
     def run_downloader(self):
         """This is the actual download thread.
@@ -329,12 +542,13 @@ class RemoteDownloader(DDBObject):
 
             self.url = url
             logging.debug("downloading url %s", self.url)
-            c = command.StartNewDownloadCommand(RemoteDownloader.dldaemon,
-                                                self.url, self.dlid,
-                                                self.contentType,
-                                                self.channelName)
-            c.send()
-            _downloads[self.dlid] = self
+            args = dict(url=self.url, content_type=self.contentType,
+                        channel_name=self.channelName,
+                        subcmd=app.download_state_manager.RESUME_NEW)
+            app.download_state_manager.add_download(self.dlid, self)
+            app.download_state_manager.queue(self.dlid,
+                                             app.download_state_manager.RESUME,
+                                             args)
             self.status["state"] = u"downloading"
         else:
             self.status["state"] = u'failed'
@@ -344,10 +558,11 @@ class RemoteDownloader(DDBObject):
 
     def pause(self):
         """Pauses the download."""
-        if _downloads.has_key(self.dlid):
-            c = command.PauseDownloadCommand(RemoteDownloader.dldaemon,
-                                             self.dlid)
-            c.send()
+        if app.download_state_manager.get_download(self.dlid):
+            args = dict(upload=False)
+            app.download_state_manager.queue(self.dlid,
+                                             app.download_state_manager.PAUSE,
+                                             args)
         self.before_changing_status()
         self.status["state"] = u"paused"
         self.after_changing_status()
@@ -359,11 +574,13 @@ class RemoteDownloader(DDBObject):
         """
         if self.get_state() in [u'downloading', u'uploading', u'paused',
                                 u'offline']:
-            if _downloads.has_key(self.dlid):
-                c = command.StopDownloadCommand(RemoteDownloader.dldaemon,
-                                                self.dlid, delete)
-                c.send()
-                del _downloads[self.dlid]
+            if app.download_state_manager.get_download(self.dlid):
+                args = dict(upload=False, delete=delete)
+                app.download_state_manager.queue(
+                    self.dlid,
+                    app.download_state_manager.STOP,
+                    args)
+                app.download_state_manager.delete_download(self.dlid)
 
         if delete:
             self.delete()
@@ -400,8 +617,7 @@ class RemoteDownloader(DDBObject):
         if self.get_state() == u'failed':
             # For failed downloads, don't trust the redirected URL (#14232)
             self.url = self.origURL
-            if _downloads.has_key (self.dlid):
-                del _downloads[self.dlid]
+            app.download_state_manager.delete_download(self.dlid)
             self.dlid = generate_dlid()
             self.before_changing_status()
             self.status = {}
@@ -412,17 +628,18 @@ class RemoteDownloader(DDBObject):
                 self.run_downloader()
             self.signal_change()
         elif self.get_state() in (u'stopped', u'paused', u'offline'):
-            if _downloads.has_key(self.dlid):
-                c = command.StartDownloadCommand(RemoteDownloader.dldaemon,
-                                                 self.dlid)
-                c.send()
-
+            if app.download_state_manager.get_download(self.dlid):
+                args = dict(subcmd=app.download_state_manager.RESUME_EXISTING)
+                app.download_state_manager.queue(
+                    self.dlid,
+                    app.download_state_manager.RESUME,
+                    args)
             self.status['state'] = u'downloading'
             self.restart()
             self.signal_change()
 
     def migrate(self, directory):
-        if _downloads.has_key(self.dlid):
+        if app.download_state_manager.get_download(self.dlid):
             c = command.MigrateDownloadCommand(RemoteDownloader.dldaemon,
                                                self.dlid, directory)
             c.send()
@@ -497,11 +714,9 @@ class RemoteDownloader(DDBObject):
     def remove(self):
         """Removes downloader from the database and deletes the file.
         """
-        global total_down_rate
-        global total_up_rate
         rates = self._get_rates()
-        total_down_rate -= rates[0]
-        total_up_rate -= rates[1]
+        app.download_state_manager.total_down_rate -= rates[0]
+        app.download_state_manager.total_up_rate -= rates[1]
         self.stop(self.delete_files)
         DDBObject.remove(self)
 
@@ -638,6 +853,8 @@ class RemoteDownloader(DDBObject):
         return self.status.get('filename', FilenameType(''))
 
     def setup_restored(self):
+        self.status_updates_frozen = False
+        self.last_update = time.time()
         self._save_later_dc = None
         self._update_retry_time_dc = None
         self.delete_files = True
@@ -667,7 +884,7 @@ class RemoteDownloader(DDBObject):
     def restart_on_startup_if_needed(self):
         if not self.id_exists():
             return
-        if _downloads.has_key(self.dlid):
+        if app.download_state_manager.get_download(self.dlid):
             # something has caused us to restart already, (for
             # example, the user selects "resume seeding").  squelch
             # any automatic behaviour (#12462)
@@ -689,16 +906,17 @@ class RemoteDownloader(DDBObject):
             else:
                 self.run_downloader()
         else:
-            _downloads[self.dlid] = self
+            app.download_state_manager.add_download(self.dlid, self)
             dler_status = self.status
             # FIXME: not sure why this is necessary
             if self.contentType ==  u'application/x-magnet':
                 dler_status['url'] = self.url
             dler_status['metainfo'] = self.metainfo
-
-            c = command.RestoreDownloaderCommand(RemoteDownloader.dldaemon,
-                                                 dler_status)
-            c.send()
+            args = dict(downloader=dler_status)
+            app.download_state_manager.queue(
+                self.dlid,
+                app.download_state_manager.RESTORE,
+                args)
             self.before_changing_status()
             self.status['state'] = u'downloading'
             self.after_changing_status()
@@ -721,10 +939,11 @@ class RemoteDownloader(DDBObject):
                          self.get_state())
             return
         self.manualUpload = True
-        if _downloads.has_key(self.dlid):
-            c = command.StartDownloadCommand(RemoteDownloader.dldaemon,
-                                             self.dlid)
-            c.send()
+        if app.download_state_manager.get_download(self.dlid):
+            args = dict(subcmd=app.download_state_manager.RESUME_EXISTING)
+            app.download_state_manager.queue(self.dlid,
+                                             app.download_state_manager.RESUME,
+                                             args)
         else:
             self.before_changing_status()
             self.status['state'] = u'uploading'
@@ -736,11 +955,11 @@ class RemoteDownloader(DDBObject):
         """
         Stop uploading/seeding and set status as "finished".
         """
-        if _downloads.has_key(self.dlid):
-            c = command.StopUploadCommand(RemoteDownloader.dldaemon,
-                                          self.dlid)
-            c.send()
-            del _downloads[self.dlid]
+        if app.download_state_manager.get_download(self.dlid):
+            args = dict(upload=True)
+            app.download_state_manager.queue(self.dlid,
+                app.download_state_manager.STOP, args)
+            app.download_state_manager.delete_download(self.dlid)
         self.before_changing_status()
         self.status["state"] = u"finished"
         self.after_changing_status()
@@ -750,11 +969,12 @@ class RemoteDownloader(DDBObject):
         """
         Stop uploading/seeding and set status as "uploading-paused".
         """
-        if _downloads.has_key(self.dlid):
-            c = command.PauseUploadCommand(RemoteDownloader.dldaemon,
-                                           self.dlid)
-            c.send()
-            del _downloads[self.dlid]
+        if app.download_state_manager.get_download(self.dlid):
+            args = dict(upload=True)
+            app.download_state_manager.queue(self.dlid,
+                app.download_state_manager.PAUSE,
+                args)
+            app.download_state_manager.delete_download(self.dlid)
         self.before_changing_status()
         self.status["state"] = u"uploading-paused"
         self.after_changing_status()
@@ -853,31 +1073,6 @@ class DownloadDaemonStarter(object):
         shutdown_downloader_objects()
         self.shutdown_callback()
         del self.shutdown_callback
-
-def init_controller():
-    """Intializes the download daemon controller.
-
-    This doesn't actually start up the downloader daemon, that's done
-    in startup_downloader.  Commands will be queued until then.
-    """
-    global daemon_starter
-    daemon_starter = DownloadDaemonStarter()
-
-def startup_downloader():
-    """Initialize the downloaders.
-
-    This method currently does 2 things.  It deletes any stale files
-    self in Incomplete Downloads, then it restarts downloads that have
-    been restored from the database.  It must be called before any
-    RemoteDownloader objects get created.
-    """
-    daemon_starter.startup()
-
-def shutdown_downloader(callback=None):
-    if daemon_starter:
-        daemon_starter.shutdown(callback)
-    elif callback:
-        callback()
 
 def lookup_downloader(url):
     try:

--- a/tv/lib/frontends/widgets/application.py
+++ b/tv/lib/frontends/widgets/application.py
@@ -1424,7 +1424,7 @@ class WidgetsMessageHandler(messages.MessageHandler):
         if share.mount:
             messages.SharingEject(share).send_to_backend()
 
-    def handle_downloader_sync_command_comlete(self, message):
+    def handle_downloader_sync_command_complete(self, message):
         # This callback is to ensure that if we are in the downloads
         # resort_on_update is re-enabled.  Walk the display stack to see
         # if there is a downloading display and if there is, reset the
@@ -1433,7 +1433,7 @@ class WidgetsMessageHandler(messages.MessageHandler):
         for d in displays:
             if d.type == 'downloading':
                 d.controller.item_list.set_resort_on_update(True)
-        logging.debug('DownloadBatchCommandComplete')
+        logging.debug('DownloaderSyncCommandComplete')
 
     def handle_jettison_tabs(self, message):
         typ = message.type

--- a/tv/lib/frontends/widgets/downloadscontroller.py
+++ b/tv/lib/frontends/widgets/downloadscontroller.py
@@ -36,6 +36,7 @@ from miro.frontends.widgets.itemlistwidgets import (
 from miro.frontends.widgets import itemcontextmenu
 from miro.frontends.widgets import prefpanel
 
+from miro import app
 from miro import messages
 from miro import downloader
 from miro import prefs
@@ -90,7 +91,10 @@ class DownloadsController(itemlistcontroller.ItemListController):
     #
     # When complete, the backend is expected to send a reply indicating that
     # all operations (up to this point) are complete and at that point we can
-    # re-enable the sort again.  See DownloadBatchCommandComplete.
+    # re-enable the sort again.  See DownloadSyncCommandComplete().  We hope
+    # for the best that the remote command to do whatever's needed should not
+    # take more than a second or two which should be accurate for a moderately
+    # sized download list.
     def _on_pause_all(self, widget):
         self.item_list.set_resort_on_update(False)
         messages.PauseAllDownloads().send_to_backend()
@@ -111,4 +115,5 @@ class DownloadsController(itemlistcontroller.ItemListController):
 
     def on_items_changed(self):
         self.status_toolbar.update_rates(
-            downloader.total_down_rate, downloader.total_up_rate)
+            app.download_state_manager.total_down_rate,
+            app.download_state_manager.total_up_rate)

--- a/tv/lib/messagehandler.py
+++ b/tv/lib/messagehandler.py
@@ -1393,12 +1393,12 @@ New ids: %s""", playlist_item_ids, message.item_ids)
 
     def handle_pause_all_downloads(self, message):
         """Pauses all downloading and uploading items"""
+        app.download_state_manager.set_bulk_mode()
         for item_ in item.Item.downloading_view():
             if item_.is_uploading():
                 item_.pause_upload()
             else:
                 item_.pause()
-        downloader.sync_downloader_commands() 
 
     def handle_pause_download(self, message):
         try:
@@ -1407,16 +1407,15 @@ New ids: %s""", playlist_item_ids, message.item_ids)
             logging.warn("PauseDownload: Item not found -- %s", message.id)
         else:
             item_.pause()
-        downloader.sync_downloader_commands()
 
     def handle_resume_all_downloads(self, message):
         """Resumes downloading and uploading items"""
+        app.download_state_manager.set_bulk_mode()
         for item_ in item.Item.paused_view():
             if item_.is_uploading_paused():
                 item_.start_upload()
             else:
                 item_.resume()
-        downloader.sync_downloader_commands()
 
     def handle_resume_download(self, message):
         try:
@@ -1427,6 +1426,7 @@ New ids: %s""", playlist_item_ids, message.item_ids)
             item_.resume()
 
     def handle_cancel_all_downloads(self, message):
+        app.download_state_manager.set_bulk_mode()
         for item_ in item.Item.download_tab_view():
             if item_.is_uploading() or item_.is_uploading_paused():
                 item_.stop_upload()

--- a/tv/lib/messages.py
+++ b/tv/lib/messages.py
@@ -907,7 +907,6 @@ class ForceFeedparserProcessing(BackendMessage):
     pass
 
 # Frontend Messages
-
 class DownloaderSyncCommandComplete(FrontendMessage):
     """Tell the frontend that the pause/resume all command are complete,
     so that we only sort once.  This saves time sorting and also prevents

--- a/tv/lib/startup.py
+++ b/tv/lib/startup.py
@@ -322,7 +322,9 @@ def finish_startup(obj, thread):
     setup_theme()
     install_message_handler()
     itemsource.setup_handlers()
-    downloader.init_controller()
+
+    app.download_state_manager = downloader.DownloadStateManager()
+    app.download_state_manager.init_controller()
 
     # Call this late, after the message handlers have been installed.
     app.sharing_tracker = sharing.SharingTracker()
@@ -423,7 +425,7 @@ def on_frontend_started():
     autoupdate.check_for_updates()
     yield None
     # Delay running high CPU/IO operations for a bit
-    eventloop.add_timeout(5, downloader.startup_downloader,
+    eventloop.add_timeout(5, app.download_state_manager.startup_downloader,
             "start downloader daemon")
     eventloop.add_timeout(10, workerprocess.startup,
             "start worker process")

--- a/tv/lib/test/downloadertest.py
+++ b/tv/lib/test/downloadertest.py
@@ -21,14 +21,15 @@ class DownloaderTest(EventLoopTest):
         app.config.set(prefs.MOVIES_DIRECTORY, self.movies_dir)
 
         # initialize and start the downloader after fixing the MOVIES_DIRECTORY
-        downloader.init_controller()
-        downloader.startup_downloader()
+        # no need to create DownloadStateManager() object.  Aleady created.
+        app.download_state_manager.init_controller()
+        app.download_state_manager.startup_downloader()
 
     def tearDown(self):
-        downloader.shutdown_downloader(
+        app.download_state_manager.shutdown_downloader(
                 lambda: self.stopEventLoop(abnormal=False))
         self.runEventLoop()
-        downloader.daemon_starter = None
+        app.download_state_manager.daemon_starter = None
         EventLoopTest.tearDown(self)
 
     def run_eventloop_until_items(self):

--- a/tv/lib/test/framework.py
+++ b/tv/lib/test/framework.py
@@ -304,6 +304,10 @@ class MiroTestCase(unittest.TestCase):
         # app.extension_manager.load_extension()
         app.extension_manager = extensionmanager.ExtensionManager(
                 [self.tempdir], [])
+        # Create a download state object (but don't start the downloader
+        # for the individual test unless necessary.  In this case we override
+        # the class to run the downloader).
+        app.download_state_manager = downloader.DownloadStateManager()
 
     def set_temp_support_directory(self):
         self.sandbox_support_directory = os.path.join(self.tempdir, 'support')

